Portal/src/Datahub.Infrastructure/Services/Notebooks/DatabricksApiService.cs

```diff
@@ -0,0 +1,170 @@
+    {
+        var workspaceDatabricksUrl = await GetWorkspaceDatabricksUrl(projectAcronym);
+
+        // Use the access token to call a protected web API.
+        var httpClient = _httpClientFactory.CreateClient();
+        var queryUrl = $"{workspaceDatabricksUrl}/api/2.0/repos";
+        httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", accessToken);
+        using var response = await httpClient.GetAsync(queryUrl);
+
+        response.EnsureSuccessStatusCode();
+        var contentString = await response.Content.ReadAsStringAsync();
+        var content = JsonSerializer.Deserialize<JsonObject>(contentString);
+
+        var existingRepositoryPreferences = await GetWorkspaceRepositoryVisibilitiesAsync(projectAcronym);
+
+        var results = content?["repos"]?
+            .AsArray()
+            .Select(node => new RepositoryInfoDto(node))
+            .ToList() ?? new List<RepositoryInfoDto>();
+
+        foreach (var repositoryInfo in results)
+        {
+            if (existingRepositoryPreferences.TryGetValue(repositoryInfo.Url, out var publicPreference))
+            {
+                repositoryInfo.IsPublic = publicPreference;
+            }
+        }
+
+        return results;
+    }
+
+    public async Task<bool> UpdateWorkspaceRepository(string projectAcronym, RepositoryInfoDto repositoryInfoDto)
+    {
+        await using var dbContext = await _dbContextFactory.CreateDbContextAsync();
+
+        var project = await dbContext.Projects
+            .FirstOrDefaultAsync(p => p.Project_Acronym_CD == projectAcronym);
+
+        if (project == null)
+        {
+            _logger.LogError("Project with acronym {ProjectAcronym} not found", projectAcronym);
+            return false;
+        }
+
+        var projectRepository = await dbContext.ProjectRepositories
+            .FirstOrDefaultAsync(pr => pr.ProjectId == project.Project_ID
+                                       && pr.RepositoryUrl == repositoryInfoDto.Url);
+
+        if (projectRepository == null)
+        {
+            projectRepository = new ProjectRepository()
+            {
+                Project = project,
+                RepositoryUrl = repositoryInfoDto.Url,
+                IsPublic = repositoryInfoDto.IsPublic,
+                Branch = repositoryInfoDto.Branch,
+                HeadCommitId = repositoryInfoDto.HeadCommitId,
+                Provider = repositoryInfoDto.Provider,
+                Path = repositoryInfoDto.Path,
+            };
+
+            dbContext.ProjectRepositories.Add(projectRepository);
+        }
+        else
+        {
+            projectRepository.IsPublic = repositoryInfoDto.IsPublic;
+            projectRepository.Branch = repositoryInfoDto.Branch;
+            projectRepository.HeadCommitId = repositoryInfoDto.HeadCommitId;
+            projectRepository.Provider = repositoryInfoDto.Provider;
+            projectRepository.Path = repositoryInfoDto.Path;
+        }
```

Comment:
I know it detects the changes, but it looks more balanced if you explicitly invoke the Update(projectRepository) here.

Reply:
Cleaned it up with 

```csharp
if (projectRepository == null)
{
    projectRepository = new ProjectRepository();
    dbContext.ProjectRepositories.Add(projectRepository);
}
    
projectRepository.Project = project;
projectRepository.RepositoryUrl = repositoryInfoDto.Url;
projectRepository.IsPublic = repositoryInfoDto.IsPublic;
projectRepository.Branch = repositoryInfoDto.Branch;
projectRepository.HeadCommitId = repositoryInfoDto.HeadCommitId;
projectRepository.Provider = repositoryInfoDto.Provider;
projectRepository.Path = repositoryInfoDto.Path;

await dbContext.SaveChangesAsync();
```